File: impls/rust/src/step4_if_fn_do.rs
use std::{cell::RefCell, collections::HashMap, io::Write, rc::Rc};

use rust2::{read_str, Env, MalError, MalFn, MalResult, MalVal, NS};

fn read(input: &str) -> Result<MalVal, MalError> {
    read_str(input)
}

fn eval_ast(ast: Rc<MalVal>, env: Rc<RefCell<Env>>) -> MalResult {
    match ast.as_ref() {
        MalVal::Symbol(symbol) => env
            .borrow()
            .get(symbol)
            .ok_or_else(|| MalError::Other(format!("'{symbol}' not found."))),
        MalVal::List(list, _) => {
            let mut buffer = Vec::new();
            for v in list {
                buffer.push(eval(v.clone(), env.clone())?);
            }
            Ok(Rc::new(MalVal::List(buffer, None)))
        }
        MalVal::Vector(vector, _) => {
            let mut buffer = Vec::new();
            for v in vector {
                buffer.push(eval(v.clone(), env.clone())?);
            }
            Ok(Rc::new(MalVal::Vector(buffer, None)))
        }
        MalVal::HashMap(hashmap, _) => {
            let mut buffer = HashMap::new();
            for (k, v) in hashmap {
                buffer.insert(k.clone(), eval(v.clone(), env.clone())?);
            }
            Ok(Rc::new(MalVal::HashMap(buffer, None)))
        }
        _ => Ok(ast),
    }
}

fn eval(ast: Rc<MalVal>, env: Rc<RefCell<Env>>) -> MalResult {
    match ast.as_ref() {
        MalVal::List(list, _) if list.is_empty() => Ok(ast),
        MalVal::List(list, _) => {
            if let MalVal::Symbol(symbol) = list[0].as_ref() {
                match symbol.as_str() {
                    "def!" => match list[1].as_ref() {
                        MalVal::Symbol(symbol) => {
                            let v = eval(list[2].clone(), env.clone())?;
                            env.borrow_mut().set(symbol.clone(), v.clone());
                            return Ok(v);
                        }
                        _ => unreachable!(),
                    },
                    "let*" => {
                        let n_env = Rc::new(RefCell::new(Env::new(env)));
                        match list[1].as_ref() {
                            MalVal::List(binds, _) | MalVal::Vector(binds, _) => {
                                let mut iter = binds.iter();
                                while let Some(v) = iter.next() {
                                    match v.as_ref() {
                                        MalVal::Symbol(symbol) => {
                                            let value =
                                                eval(iter.next().unwrap().clone(), n_env.clone())?;
                                            n_env.borrow_mut().set(symbol.clone(), value);
                                        }
                                        _ => unreachable!(),
                                    }
                                }
                            }
                            _ => unreachable!(),
                        }
                        return eval(list[2].clone(), n_env);
                    }
                    "do" => {
                        let mut value = eval(list[1].clone(), env.clone())?;
                        for v in &list[2..] {
                            value = eval(v.clone(), env.clone())?;
                        }
                        return Ok(value);
                    }
                    "if" => {
                        let condition = eval(list[1].clone(), env.clone())?;
                        return match condition.as_ref() {
                            MalVal::Nil | MalVal::Bool(false) => {
                                if list.len() > 3 {
                                    eval(list[3].clone(), env)
                                } else {
                                    Ok(Rc::new(MalVal::Nil))
                                }
                            }
                            _ => eval(list[2].clone(), env),
                        };
                    }
                    "fn*" => {
                        let binds = match list[1].as_ref() {
                            MalVal::List(list, _) | MalVal::Vector(list, _) => list
                                .iter()
                                .map(|v| match v.as_ref() {
                                    MalVal::Symbol(symbol) => symbol.to_string(),
                                    _ => unreachable!(),
                                })
                                .collect::<Vec<_>>(),
                            _ => unreachable!(),
                        };
                        let body = list[2].clone();
                        return Ok(Rc::new(MalVal::Fn(
                            Rc::new(MalFn::custom_func(body, binds, env, eval)),
                            None,
                        )));
                    }
                    _ => (),
                }
            }
            let ast = eval_ast(ast, env)?;
            match ast.as_ref() {
                MalVal::List(list, _) => match list[0].as_ref() {
                    MalVal::Fn(func, _) => match func.as_ref() {
                        MalFn::RegularFn(func) => (func)(&list[1..]),
                        MalFn::MalFunc(func) => {
                            let mut n_env = Env::new(func.env.clone());
                            n_env.bind_expr(func.params.clone(), list[1..].to_vec());
                            eval(func.ast.clone(), Rc::new(RefCell::new(n_env)))
                        }
                    },
                    _ => unreachable!(),
                },
                _ => unreachable!(),
            }
        }
        _ => eval_ast(ast, env),
    }
}

fn print(val: &MalVal) -> String {
    val.pr_str(true)
}

fn rep(input: &str, env: &Rc<RefCell<Env>>) -> String {
    match read(input) {
        Ok(ast) => match eval(Rc::new(ast), env.clone()) {
            Ok(v) => print(v.as_ref()),
            Err(e) => e.to_string(),
        },
        Err(e) => e.to_string(),
    }
}

fn main() {
    let mut env = Env::default();
    for (k, v) in NS {
        env.set(
            k.to_string(),
            Rc::new(MalVal::Fn(Rc::new(MalFn::RegularFn(Rc::new(v))), None)),
        );
    }
    let env = Rc::new(RefCell::new(env));
    rep("(def! not (fn* (a) (if a false true)))", &env);

    let mut buffer = String::new();
    loop {
        print!("user> ");
        std::io::stdout().flush().unwrap();
        let input = match std::io::stdin().read_line(&mut buffer) {
            Ok(0) => {
                println!();
                break;
            }
            Err(_) => break,
            Ok(_) => buffer.trim(),
        };
        if !input.is_empty() {
            println!("{}", rep(input, &env));
        }
        buffer.clear();
    }
}

File: impls/cpp/src/step4_if_fn_do.cpp
#include <fmt/core.h>

#include <functional>
#include <string>
#include <string_view>

#include "core.h"
#include "env.h"
#include "linenoise.h"
#include "printer.h"
#include "reader.h"

using namespace std::literals;

std::shared_ptr<MalType> Eval(const std::shared_ptr<MalType>& ast,
                              std::shared_ptr<Env> env);
std::shared_ptr<MalType> EvalAst(const std::shared_ptr<MalType>& ast,
                                 std::shared_ptr<Env> env) {
    if (auto s = std::dynamic_pointer_cast<Symbol>(ast)) {
        return env->Get(**s);
    } else if (auto v = std::dynamic_pointer_cast<Vector>(ast)) {
        auto vector = std::make_shared<Vector>();
        for (auto& item : **v) (*vector)->emplace_back(Eval(item, env));
        return vector;
    } else if (auto l = std::dynamic_pointer_cast<List>(ast)) {
        auto list = std::make_shared<List>();
        for (auto& v : **l) (*list)->emplace_back(Eval(v, env));
        return list;
    } else if (auto h = std::dynamic_pointer_cast<HashMap>(ast)) {
        auto map = std::make_shared<HashMap>();
        for (auto& [k, v] : **h) (*map)->emplace(k, Eval(v, env));
        return map;
    }
    return ast;
}

auto Read(std::string_view str) { return ReadStr(str); }

std::shared_ptr<MalType> Eval(const std::shared_ptr<MalType>& ast,
                              std::shared_ptr<Env> env) {
    if (auto l = std::dynamic_pointer_cast<List>(ast)) {
        if ((*l)->empty()) return ast;
        auto it = (*l)->begin();
        if (auto s = std::dynamic_pointer_cast<Symbol>(*it)) {
            if (const auto& action = **s; action == "def!") {
                auto symbol = **std::dynamic_pointer_cast<Symbol>(*(++it));
                auto value = Eval(*(++it), env);
                env->Set(std::move(symbol), value);
                return value;
            } else if (action == "let*") {
                auto new_env = std::make_shared<Env>(env);
                auto bindings = std::dynamic_pointer_cast<Sequence>(*(++it));
                for (auto it = (*bindings)->begin();
                     it + 1 < (*bindings)->end(); it += 2) {
                    auto symbol = **std::dynamic_pointer_cast<Symbol>(*it);
                    auto value = Eval(*(it + 1), new_env);
                    new_env->Set(std::move(symbol), std::move(value));
                }
                return Eval(*(++it), new_env);
            } else if (action == "do") {
                std::shared_ptr<MalType> res;
                for (++it; it != (*l)->end(); it++) res = Eval(*it, env);
                return res;
            } else if (action == "if") {
                auto condition = Eval(*(++it), env);
                if (std::dynamic_pointer_cast<False>(condition) ||
                    std::dynamic_pointer_cast<Nil>(condition)) {
                    if ((*l)->size() < 4) return MalType::Nil;
                    return Eval(*(it + 2), env);
                } else {
                    return Eval(*(++it), env);
                }
            } else if (action == "fn*") {
                auto binds_list =
                    std::dynamic_pointer_cast<Sequence>((*l)->at(1));
                auto body = (*l)->at(2);
                std::vector<std::string> binds((*binds_list)->size());
                std::transform(
                    (*binds_list)->begin(), (*binds_list)->end(), binds.begin(),
                    [](std::shared_ptr<MalType> arg) -> std::string {
                        return **std::dynamic_pointer_cast<Symbol>(arg);
                    });
                auto f = [body = std::move(body), env,
                          binds = std::move(binds)](
                             std::span<std::shared_ptr<MalType>> args) {
                    auto new_env = std::make_shared<Env>(
                        std::span{binds.begin(), binds.end()}, args, env);
                    return Eval(body, new_env);
                };
                return std::make_shared<BaseFunc>(f);
            }
        }
        auto list = std::dynamic_pointer_cast<List>(EvalAst(l, env));
        it = (*list)->begin();
        auto func = std::dynamic_pointer_cast<MalFunc>(*(it++));
        return (*func)(std::span<std::shared_ptr<MalType>>(it, (*list)->end()));
    }
    return EvalAst(ast, env);
}

std::string Print(auto ast) { return PrStr(ast, true); }
std::string Rep(std::string_view str, std::shared_ptr<Env> env) {
    try {
        return Print(Eval(Read(str), env));
    } catch (std::string_view err) {
        fmt::print("{}", err);
        return "";
    } catch (std::string err) {
        fmt::print("{}", err);
        return "";
    }
}

int main() {
    linenoiseHistorySetMaxLen(30);
    linenoiseHistoryLoad("history.txt");

    auto env = std::make_shared<Env>();

    for (auto& [k, v] : getNS()) {
        auto f = std::make_shared<BaseFunc>(v);
        env->Set(std::string{k}, f);
    }

    Rep("(def! not (fn* (a) (if a false true)))", env);

    char* line;
    while ((line = linenoise("user> ")) != NULL) {
        if (line[0] == '\0') continue;
        linenoiseHistoryAdd(line);
        try {
            fmt::print("{}\n", Rep(line, env));
        } catch (std::nullptr_t) {
            linenoiseFree(line);
            continue;
        }
        linenoiseFree(line);
    }
    linenoiseHistorySave("history.txt");
    return 0;
}
